azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -5,150 +5,166 @@
-
-    def _generic_handler_setter(self, handler_name, new_handler):
-        """Set a handler"""
-        curr_handler = getattr(self, handler_name)
-        if new_handler is not None and curr_handler is None:
-            # Create task, set handler
-            logger.debug("Creating new handler task for handler: {}".format(handler_name))
-            setattr(self, handler_name, new_handler)
-            self._add_handler_task(handler_name)
-        elif new_handler is None and curr_handler is not None:
-            # Cancel task, remove handler
-            logger.debug("Removing handler task for handler: {}".format(handler_name))
-            self._remove_handler_task(handler_name)
-            setattr(self, handler_name, new_handler)
+            coro = self._inbox_handler_runner(inbox, handler_name)
         else:
-            # Update handler, no need to change tasks
-            logger.debug("Updating handler for handler: {}".format(handler_name))
-            setattr(self, handler_name, new_handler)
-
-    @property
-    def on_message_received(self):
-        return self._on_message_received
-
-    @on_message_received.setter
-    def on_message_received(self, value):
-        self._generic_handler_setter(MESSAGE, value)
-
-    @property
-    def on_method_request_received(self):
-        return self._on_method_request_received
-
-    @on_method_request_received.setter
-    def on_method_request_received(self, value):
-        self._generic_handler_setter(METHOD, value)
-
-    @property
-    def on_twin_desired_properties_patch_received(self):
-        return self._on_twin_desired_properties_patch_received
-
-    @on_twin_desired_properties_patch_received.setter
-    def on_twin_desired_properties_patch_received(self, value):
-        self._generic_handler_setter(TWIN_DP_PATCH, value)
+            coro = self._event_handler_runner(handler_name)
+        future = asyncio.run_coroutine_threadsafe(coro, RUNNER_LOOP)
+
+        # Define a callback for the future (in order to handle any errors)
+        def _handler_runner_callback(completed_future):
+            try:
+                e = completed_future.exception()
+            except asyncio.CancelledError:
+                # TODO: is this the right error?
+                new_err = HandlerManagerException(
+                    message="HANDLER RUNNER ({}): Task unexpectedly ended in cancellation".format(
+                        handler_name
+                    )
+                )
+                handle_exceptions.handle_background_exception(new_err)
+            else:
+                if e:
+                    new_err = HandlerManagerException(
+                        message="HANDLER RUNNER ({}): Unexpected error during task".format(
+                            handler_name
+                        ),
+                        cause=e,
+                    )
+                    handle_exceptions.handle_background_exception(new_err)
```

Comment:
>handle_exceptions.handle_background_exception(new_err) [](start = 20, length = 54)

This is a very underwhelming response to a very serious error.  if the handler runner stops running, then the handler will never be called again.  App developers have no way to detect this condition unless they happen to notice the background exception or the lack of handlers being called.  This feels fatal to me.  It would be better to crash spectacularly or try to restart the handler runner.